File: fsharp/perfect-numbers/PerfectNumbers.fs
module PerfectNumbers

    open System

    type Classification =
        | Perfect
        | Abundant
        | Deficient

    let private accumulateFactors (n:int) (accumulator:int) (number:int):int =
        match n % number with
        | 0 ->
            match number = n / number with
            | true -> accumulator + number
            | false -> accumulator + number + (n / number)
        | _ -> accumulator

    let private aliquotSum (n:int):int =
        match n with
        | 1 -> 0
        | _ ->
            let ceiling:int =
                n
                |> double
                |> Math.Sqrt
                |> int
            seq{2..ceiling}
            |> Seq.fold (accumulateFactors n) 1


    let classify (n:int):Classification option =
        match n with
        | n when n > 0 ->
            match n |> aliquotSum |> n.CompareTo |> Math.Sign with
            | -1 -> Abundant
            | 0 -> Perfect
            | 1 -> Deficient
            |> Some
        | _ -> None

File: solutions/scala/perfect-numbers/1/src/main/scala/PerfectNumbers.scala
enum NumberType
{
    case Perfect
    case Abundant
    case Deficient
}

object PerfectNumbers
{

    private def accumulateFactors(n:Int, accumulator:Int, number:Int):Int =
    {
        n % number match
        {
            case 0 =>
                number == n / number match
                case true => accumulator + number
                case false => accumulator + number + (n / number)
            case _ => accumulator
        }
    }

    private def aliquotSum(n:Int):Int =
    {
        n match
        {
            case 1 => 0
            case _ =>
                val ceiling:Int = (Math.sqrt(n.toDouble)).toInt
                (2 to ceiling).foldLeft(1)((accumulator:Int, number:Int) => accumulateFactors(n, accumulator, number))
        }
    }


    def classify(n:Int):Either[String, NumberType] =
    {
        n match
        {
            case n if n > 0 =>
                val numberType:NumberType =
                    Math.signum(n.compareTo(aliquotSum(n))) match
                    {
                        case -1 => NumberType.Abundant
                        case 0 => NumberType.Perfect
                        case 1 => NumberType.Deficient
                    }
                Right(numberType)
            case _ => Left("Classification is only possible for natural numbers.")
        }
    }

}